### PR TITLE
COPY support on master node for append partitioned relations

### PR DESCRIPTION
**Pull request by metdos**
cr: @marcocitus

### CODE REVIEW

File: src/backend/distributed/commands/multi_copy.c [hunk 1]
```diff
@@ -740,7 +779,7 @@ OpenCopyTransactions(CopyStmt *copyStatement, ShardConnections *shardConnections
 	List *connectionList = NULL;
 
 	MemoryContext localContext = AllocSetContextCreate(CurrentMemoryContext,
-													   "OpenCopyTransactions",
+													   "Open Copy Transactions",
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: Why this change? It seems we usually have the function name here.

File: src/backend/distributed/commands/multi_copy.c [hunk 2]
```diff
@@ -1195,6 +1234,282 @@ AppendCopyBinaryFooters(CopyOutState footerOutputState)
...
+		/* roll back all transactions */
+		EndRemoteCopy(shardConnections->connectionList, false);
+		AbortRemoteTransactions(shardConnections->connectionList);
+		CloseConnections(shardConnections->connectionList);
+
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	if (completionTag != NULL)
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: Does this need to be below the PG_TRY?

File: src/backend/distributed/commands/multi_copy.c [hunk 3]
```diff
@@ -1195,6 +1234,282 @@ AppendCopyBinaryFooters(CopyOutState footerOutputState)
...
+	{
+		char *relationName = get_rel_name(relationId);
+
+		Oid schemaId = get_rel_namespace(relationId);
+		char *schemaName = get_namespace_name(schemaId);
+
+		ShardInterval *shardInterval = LoadShardInterval(shardConnections->shardId);
+		List *shardIntervalList = list_make1(shardInterval);
+
+		DropShards(relationId, schemaName, relationName, shardIntervalList);
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: Hmz, this feels a little hacky. Could we wait until we have at least one row before creating new shards?

File: src/backend/distributed/commands/multi_copy.c [hunk 4]
```diff
@@ -1195,6 +1234,282 @@ AppendCopyBinaryFooters(CopyOutState footerOutputState)
...
+	}
+}
+
+
+/*
+ * UpdateShardStatistics updates metadata for the given shard id.
+ */
+static void
+UpdateShardStatistics(Oid relationId, int64 shardId)
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: Are there other places where we do this? Is it possible to use the same logic there? Also wondering whether this is the right place.

File: src/test/regress/output/multi_copy.source [hunk 5]
```diff
@@ -225,3 +225,86 @@ SELECT count(*) FROM customer_copy_range;
...
+        l_linestatus char(1) not null,
+        l_shipdate date not null,
+        l_commitdate date not null,
+        l_receiptdate date not null,
+        l_shipinstruct char(25) not null,
+        l_shipmode char(10) not null,
+        l_comment varchar(44) not null,
+        PRIMARY KEY(l_orderkey, l_linenumber) );
+SELECT master_create_distributed_table('lineitem_copy_append', 'l_orderkey', 'append');
+WARNING:  table "lineitem_copy_append" has a unique constraint
```
Comment (marcocitus) [inline src/test/regress/output/multi_copy.source]: Any reason to keep the primary key in there?

File: src/backend/distributed/commands/multi_copy.c (hunk 1 above)
Comment (metdos) [inline src/backend/distributed/commands/multi_copy.c]: I memorized this line
TaskTrackerContext = AllocSetContextCreate(TopMemoryContext, "Task Tracker",

but it seems that without spaces are more common. I'm taking back the change.

File: src/backend/distributed/commands/multi_copy.c [hunk 6]
```diff
@@ -1195,6 +1233,195 @@ AppendCopyBinaryFooters(CopyOutState footerOutputState)
...
+			{
+				MemoryContextSwitchTo(oldContext);
+				break;
+			}
+
+			CHECK_FOR_INTERRUPTS();
+
+			MemoryContextSwitchTo(oldContext);
+
+			if (copiedDataSizeInBytes == 0)
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: It's not immediately obvious what this means. Maybe add a comment.

File: src/backend/distributed/commands/multi_copy.c [hunk 7]
```diff
@@ -1195,6 +1233,195 @@ AppendCopyBinaryFooters(CopyOutState footerOutputState)
...
+			/* replicate row to shard placements */
+			resetStringInfo(copyOutState->fe_msgbuf);
+			AppendCopyRowData(columnValues, columnNulls, tupleDescriptor,
+							  copyOutState, columnOutputFunctions);
+			SendCopyDataToAll(copyOutState->fe_msgbuf, shardConnections->connectionList);
+
+			messageBufferSize = copyOutState->fe_msgbuf->len;
+			copiedDataSizeInBytes = copiedDataSizeInBytes + messageBufferSize;
+
+			if (copiedDataSizeInBytes > shardMaxSizeInBytes)
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: Maybe also add a comment here saying what we're trying to achieve by doing this.

File: src/backend/distributed/commands/multi_copy.c [hunk 8]
```diff
@@ -1195,6 +1233,195 @@ AppendCopyBinaryFooters(CopyOutState footerOutputState)
...
+	TupleDesc tupleDescriptor = RelationGetDescr(distributedRelation);
+	uint32 columnCount = tupleDescriptor->natts;
+	Datum *columnValues = palloc0(columnCount * sizeof(Datum));
+	bool *columnNulls = palloc0(columnCount * sizeof(bool));
+
+	EState *executorState = CreateExecutorState();
+	MemoryContext executorTupleContext = GetPerTupleMemoryContext(executorState);
+	ExprContext *executorExpressionContext = GetPerTupleExprContext(executorState);
+
+	ShardConnections *shardConnections =
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: We should maybe add the note on PG_TRY that we have in the other function here as well.

File: src/backend/distributed/commands/multi_copy.c [hunk 9]
```diff
@@ -1195,6 +1233,195 @@ AppendCopyBinaryFooters(CopyOutState footerOutputState)
...
+
+			if (!nextRowFound)
+			{
+				MemoryContextSwitchTo(oldContext);
+				break;
+			}
+
+			CHECK_FOR_INTERRUPTS();
+
+			MemoryContextSwitchTo(oldContext);
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: Hmz, are we leaking memory into the outer context in the append functions? I realize this might be wrong in CopyToHashOrRangePartitionedTable as well.

File: src/backend/distributed/commands/multi_copy.c [hunk 10]
```diff
@@ -1195,6 +1233,195 @@ AppendCopyBinaryFooters(CopyOutState footerOutputState)
...
+			processedRowCount += 1;
+		}
+
+		/* all lines have been copied, stop showing line number in errors */
+		error_context_stack = errorCallback.previous;
+
+		/* check for cancellation one last time before committing */
+		CHECK_FOR_INTERRUPTS();
+
+		if (copiedDataSizeInBytes > 0)
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: and again :)

File: src/include/distributed/master_protocol.h [hunk 11]
```diff
@@ -85,6 +85,7 @@ extern void CheckDistributedTable(Oid relationId);
 extern void CreateShardPlacements(int64 shardId, List *ddlEventList,
 								  List *workerNodeList, int workerStartIndex,
 								  int replicationFactor);
+extern float4 UpdateShardStatistics(Oid relationId, int64 shardId);
```
Comment (marcocitus) [inline src/include/distributed/master_protocol.h]: The return value feels pretty arbitrary. I'd feel slightly more comfortable returning the size or a statistics struct, and leave arbitrary calculations like fill factor to the caller.

File: src/test/regress/input/multi_copy.source [hunk 12]
```diff
@@ -188,3 +189,70 @@ FROM customer_copy_range WHERE c_custkey <= 500;
...
+SELECT master_create_distributed_table('lineitem_copy_append', 'l_orderkey', 'append');
+
+-- Test multiple shard creation
+SET citus.shard_max_size TO '256kB';
+
+COPY lineitem_copy_append FROM '@abs_srcdir@/data/lineitem.1.data' with delimiter '|';
+
+SELECT count(*) FROM pg_dist_shard WHERE logicalrelid = 'lineitem_copy_append'::regclass;
+
+SET citus.shard_max_size TO DEFAULT;
```
Comment (marcocitus) [inline src/test/regress/input/multi_copy.source]: Seems redundant, pg_regress opens up a new session for every test file.

File: src/test/regress/output/multi_copy.source [hunk 13]
```diff
@@ -225,3 +225,82 @@ SELECT count(*) FROM customer_copy_range;
...
+     0
+(1 row)
+
+-- Test proper copy
+COPY customer_copy_append(c_custkey, c_name) FROM STDIN WITH (FORMAT 'csv');
+-- Check whether data was copied
+SELECT count(*) FROM customer_copy_append;
+ count 
+-------
+     2
```
Comment (marcocitus) [inline src/test/regress/output/multi_copy.source]: We could maybe do a basic content check here.

File: src/backend/distributed/commands/multi_copy.c [hunk 14]
```diff
@@ -1195,6 +1233,195 @@ AppendCopyBinaryFooters(CopyOutState footerOutputState)
 }
 
 
+/*
+ * CopyToAppendPartitionedTable implements the COPY table_name FROM ... for
+ * append-partitioned tables.
+ */
+static void
+CopyToAppendPartitionedTable(CopyStmt *copyStatement, char *completionTag)
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: I think it would be better to put this next to CopyToHashOrRangePartitionedTable

File: src/backend/distributed/commands/multi_copy.c [hunk 15]
```diff
@@ -233,12 +223,60 @@ CitusCopyFrom(CopyStmt *copyStatement, char *completionTag)
 
 	partitionColumn = PartitionColumn(tableId, 0);
 	partitionMethod = PartitionMethod(tableId);
-	if (partitionMethod != DISTRIBUTE_BY_RANGE && partitionMethod != DISTRIBUTE_BY_HASH)
+	if (partitionMethod == DISTRIBUTE_BY_HASH || partitionMethod == DISTRIBUTE_BY_RANGE)
+	{
+		CopyToHashOrRangePartitionedTable(copyStatement, completionTag);
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: Maybe CopyToExistingShards vs CopyToNewShards.

File: src/backend/distributed/utils/multi_transaction.c [hunk 16]
```diff
@@ -135,10 +135,11 @@ AbortRemoteTransactions(List *connectionList)
 
 /*
  * CommitRemoteTransactions commits all transactions on connections in connectionList.
- * On failure, it reports a warning and continues committing all of them.
+ * If stopOnFailure is true, then CommitRemoteTransactions reports an error on
+ * failure, otherwise it reports a warning.
```
Comment (marcocitus) [inline src/backend/distributed/utils/multi_transaction.c]: Maybe add why we would want stopOnFailure to be true or false.

File: src/backend/distributed/commands/multi_copy.c (hunk 8 above)
Comment (metdos) [inline src/backend/distributed/commands/multi_copy.c]: Could you explain this a bit more? I didn't understand what note to add to where.

File: src/backend/distributed/commands/multi_copy.c (hunk 9 above)
Comment (metdos) [inline src/backend/distributed/commands/multi_copy.c]: AppendCopyRowData() uses copyOutState->rowcontext which is set above.

EState *executorState = CreateExecutorState();
MemoryContext executorTupleContext = GetPerTupleMemoryContext(executorState);
....
copyOutState->rowcontext = executorTupleContext;

We reset it for every tuple;
ResetPerTupleExprContext(executorState);
oldContext = MemoryContextSwitchTo(executorTupleContext);

Did you mean something different?

File: src/backend/distributed/commands/multi_copy.c (hunk 8 above)
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: Where we assign the shard connections hash in CopyToHashOrRangePartitionedTable there's a note about PG_TRY subtleties because that variable is used in the PG_CATCH (which wouldn't work if it was assigned within PG_TRY), it might be useful to add the same note here or refer to the other one.

File: src/backend/distributed/commands/multi_copy.c (hunk 9 above)
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: Could we use rowcontext in AppendCopyBinaryHeaders/Footers as well?

**Comment by metdos** [comment]
Hey @marcocitus,
I addressed your feedback above. Note that I will squeeze commits before the merge.

File: src/backend/distributed/commands/multi_copy.c [hunk 17]
```diff
@@ -233,12 +222,60 @@ CitusCopyFrom(CopyStmt *copyStatement, char *completionTag)
...
-						errmsg("COPY is only supported for hash- and "
-							   "range-partitioned tables")));
+						errmsg("unsupported partition method")));
 	}
+}
+
+
+/*
+ * CopyToExistingShards implements the COPY table_name FROM ... for hash or
+ * range-partitioned tables where there are already shards to copy rows.
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: very minor: into which to copy rows.

File: src/backend/distributed/commands/multi_copy.c [hunk 18]
```diff
@@ -1190,8 +1383,69 @@ void
...
+	shardConnections->shardId = shardId;
+
+	list_free_deep(shardConnections->connectionList);
+	shardConnections->connectionList = NIL;
+
+	/* connect to shards placements and start transactions */
+	OpenCopyTransactions(copyStatement, shardConnections);
+
+	/* send binary headers to shard placements */
+	resetStringInfo(copyOutState->fe_msgbuf);
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: What do you think of moving these 3 lines into the while loop? I put them there for hash-partitioned tables partly to reduce the signature of the function and partly for subjective aesthetics (I like the symmetry between the code that sends headers, rows, footers). Alternatively, maybe we could wrap them in a function.

File: src/backend/distributed/commands/multi_copy.c [hunk 19]
```diff
@@ -462,6 +499,160 @@ CitusCopyFrom(CopyStmt *copyStatement, char *completionTag)
...
+			}
+
+			processedRowCount += 1;
+		}
+
+		/* all lines have been copied, stop showing line number in errors */
+		error_context_stack = errorCallback.previous;
+
+		/* check for cancellation one last time before committing */
+		CHECK_FOR_INTERRUPTS();
```
Comment (marcocitus) [inline src/backend/distributed/commands/multi_copy.c]: FinalizeCopyToNewShard might actually take a while. It might be better to check afterwards.

File: src/backend/distributed/commands/multi_copy.c (hunk 18 above)
Comment (metdos) [inline src/backend/distributed/commands/multi_copy.c]: I wrapped header and footer parts as separate functions.

I also think about putting row copy part into a function, but it becomes a 3 lines function with 6 parameters. Therefore, I leave it as it is. Function would be like following:

```sql
static void
SendCopyRow(CopyOutState copyOutState, List *connectionList, Datum *columnValues,
			bool *columnNulls, TupleDesc tupleDescriptor, FmgrInfo *columnOutputFunctions)
{
	resetStringInfo(copyOutState->fe_msgbuf);
	AppendCopyRowData(columnValues, columnNulls, tupleDescriptor,
					  copyOutState, columnOutputFunctions);
	SendCopyDataToAll(copyOutState->fe_msgbuf, connectionList);
}
```

File: src/backend/distributed/planner/modify_planner.c [hunk 20]
```diff
@@ -102,7 +102,6 @@ ErrorIfQueryNotSupported(Query *queryTree)
 	Oid distributedTableId = ExtractFirstDistributedTableId(queryTree);
 	uint32 rangeTableId = 1;
 	Var *partitionColumn = PartitionColumn(distributedTableId, rangeTableId);
-	char partitionMethod = PartitionMethod(distributedTableId);
```
Comment (marcocitus) [inline src/backend/distributed/planner/modify_planner.c]: Is this related to COPY?

**Comment by metdos** [comment]
Hey @marcocitus,
I addressed your latest feedback above. I also added a note for a possible refactorization.

File: src/backend/distributed/planner/modify_planner.c (hunk 20 above)
Comment (metdos) [inline src/backend/distributed/planner/modify_planner.c]: No, it is not, but this is a very small change which removes an unnecessary warning and nobody fixed for a while. Therefore, I thought that I should fix it.

**Comment by marcocitus** [comment]
What's the refactorization note? 

I think it looks good, I'll finish testing today.

**Comment by metdos** [comment]
>
> What's the refactorization note?
>
This one:

I wrapped header and footer parts as separate functions.

I also think about putting row copy part into a function, but it becomes a
3 lines function with 6 parameters. Therefore, I leave it as it is.
Function would be like following:

static void
SendCopyRow(CopyOutState copyOutState, List *connectionList, Datum
*columnValues,
            bool *columnNulls, TupleDesc tupleDescriptor, FmgrInfo
*columnOutputFunctions)
{
    resetStringInfo(copyOutState->fe_msgbuf);
    AppendCopyRowData(columnValues, columnNulls, tupleDescriptor,
                      copyOutState, columnOutputFunctions);
    SendCopyDataToAll(copyOutState->fe_msgbuf, connectionList);
}

I think it looks good, I'll finish testing today.
>
Great!

**Comment by marcocitus** [comment]
Is there a way to clean up this error message?

    WARNING:  could not connect to node localhost:9700
    DETAIL:  Client error: could not connect to server: Connection refused
            Is the server running on host "localhost" (127.0.0.1) and accepting
            TCP/IP connections on port 9700?
    
    CONTEXT:  COPY customer_reviews, line 1: "AE22YDHSBFYIP,1970-12-30,5,10,0,1551803542,"Start and Run a Coffee Bar (Start & Run a)",11611,Book,"..."
    WARNING:  could not create shard on "localhost:9700"
    CONTEXT:  COPY customer_reviews, line 1: "AE22YDHSBFYIP,1970-12-30,5,10,0,1551803542,"Start and Run a Coffee Bar (Start & Run a)",11611,Book,"..."
    ERROR:  could only create 1 of 2 of required shard replicas
    CONTEXT:  COPY customer_reviews, line 1: "AE22YDHSBFYIP,1970-12-30,5,10,0,1551803542,"Start and Run a Coffee Bar (Start & Run a)",11611,Book,"..."

**Comment by marcocitus** [comment]
Also the one that occurs if shard placement creation fails:

    postgres=# COPY customer_reviews FROM '/home/marco/citus/customer_reviews_1998.csv' WITH CSV;
    WARNING:  could not connect to node localhost:9700
    DETAIL:  Client error: could not connect to server: Connection refused
            Is the server running on host "localhost" (127.0.0.1) and accepting
            TCP/IP connections on port 9700?

    CONTEXT:  COPY customer_reviews, line 1: "AE22YDHSBFYIP,1970-12-30,5,10,0,1551803542,"Start and Run a Coffee Bar (Start & Run a)",11611,Book,"..."
    WARNING:  could not create shard on "localhost:9700"
    CONTEXT:  COPY customer_reviews, line 1: "AE22YDHSBFYIP,1970-12-30,5,10,0,1551803542,"Start and Run a Coffee Bar (Start & Run a)",11611,Book,"..."
    COPY 589859

**Comment by marcocitus** [comment]
Is there a way to recover from this issue?

    postgres=# COPY customer_reviews FROM '/home/marco/citus/customer_reviews_1998.csv' WITH CSV;
    WARNING:  could not consume data from worker node
CONTEXT:  COPY customer_reviews, line 435263: "ATVPDKIKX0DER,1998-10-12,5,1,1,0738200433,"Almost a Woman",632937,Book,"Biographies & Memoirs","Ethn..."
    WARNING:  could not get statistics for shard customer_reviews_102067
    DETAIL:  Setting shard statistics to NULL
    CONTEXT:  COPY customer_reviews, line 435263: "ATVPDKIKX0DER,1998-10-12,5,1,1,0738200433,"Almost a Woman",632937,Book,"Biographies & Memoirs","Ethn..."
    COPY 589859

**Comment by marcocitus** [comment]
I also saw this warning when killing a worker node during the copy, not sure what it means:

    postgres=# COPY customer_reviews FROM '/home/marco/citus/customer_reviews_1998.csv' WITH CSV;
    WARNING:  could not consume data from worker node
    COPY 589859

**Comment by marcocitus** [comment]
>  I also think about putting row copy part into a function, but it becomes a
> 3 lines function with 6 parameters. 

Yeh, let's not do that.

**Comment by marcocitus** [comment]
I completed my testing. We could do with nicer error messages when there are failures. We should also add a UDF for updating shard statistics, which needs to go into the upgrade script.

I also realised the begin/commit stuff doesn't actually work for cstore_fdw tables, but that's more of a problem for hash-partitioned tables. It would also be nice to have https://github.com/citusdata/cstore_fdw/pull/89 in cstore_fdw.

**Comment by metdos** [comment]
Do you mean row numbers in the error message? I added a commit to fix it.
Do you think  we need a similar approach in hash-copy?

**Comment by marcocitus** [comment]
> Do you mean row numbers in the error message?

Yes, the COPY context seems pretty noisy in the shard creation case. 

> Do you think  we need a similar approach in hash-copy?

It's less of an issue there because there are no noisy shard creation warnings, but it would be good for consistency.

**Comment by marcocitus** [comment]
:shipit: 

As part of the COPY via worker changes, we will also have a UDF for updating shard statistics, right? Once we do, it might be nice to add an errhint to the message.

File: src/backend/distributed/planner/modify_planner.c (hunk 20 above)
Comment (marcocitus) [inline src/backend/distributed/planner/modify_planner.c]: It's addressed by https://github.com/citusdata/citus/pull/450/files so better to remove it from this change.
